Sources/X509/CryptographicMessageSyntax/CMSOperations.swift

```diff
@@ -38,12 +39,17 @@ public enum CMS {
 
         // no signing time provided, sign regularly (without signedAttrs)
         let signature = try privateKey.sign(bytes: bytes, signatureAlgorithm: signatureAlgorithm)
-        let signedData = try self.generateSignedData(
+        var signedData = try self.generateSignedData(
```

Comment:
I think we should sink the generation of the signed data down into `generateSignedData`. This has two advantages.

First, it means the extension to the "with signing time" case is straightforward. That case also ends up calling into generateSignedData`, which makes our lives a lot easier.

Secondly, it means we don't have to do the transformation back and forth through `CMSContentInfo`, which makes the operation _much_ cheaper. Transforming through ASN1Any is fairly expensive and it'd be better if we set this up correctly the first time, instead of ping-ponging through.

Reply:
So this is broadly looking pretty good.

For test cases, you want to add a few new ones, rather than tweak the existing ones. You probably want the following:

1. A test case that tests that the non-signing-time path can both produce a signature, and have that signature verified, with the signature in non-detached mode. Roughly equivalent to `testSimpleSigningVerifying` but with non-detached signatures.
2. A test case like the above, but with signing time included. Roughly equivalent to `testSigningWithSigningTimeSignedAttr`.
3. A version of (1) where the signature is made non-attached, but is verified as if it were attached. This should fail.
4. A version of (2) where the signature is made non-attached, but is verified as if it were attached. This should fail.
5. 3. A version of (1) where the signature is made attached, but is verified as if it were non-attached. This should pass: all th eneeded data is present.
4. A version of (2) where the signature is made attached, but is verified as if it were nonattached. This should pass: all the needed data is present..

Reply:
There's absolutely nothing to apologise for, it's not always easy to get the time management lined up. This is the nice thing about OSS, I can be event-driven, so waiting for you to come back isn't a burden at all! I'll re-review now.